packages/protobuf/src/ast.ts

```diff
@@ -0,0 +1,279 @@
+// Copyright (c) Microsoft Corporation.
+
+import { Namespace } from "@typespec/compiler";
+
+/**
+ * This module describes an AST for Protobuf.
+ */
+
+/**
+ * A single .proto file.
+ */
+export interface ProtoFile {
+  /**
+   * The package name, if one is known.
+   */
+  package?: string;
+  /**
+   * The `option` specifiers to include in the file.
+   */
+  options: Partial<WellKnownFileOptions>;
+
+  /**
+   * Paths imported by this file.
+   */
+  imports: string[];
+
+  /**
+   * The declarations in the file.
+   *
+   * Only `service` and `message` declarations may exist at the root of the file.
+   */
+  declarations: Iterable<ProtoTopLevelDeclaration>;
+
+  /**
+   * The original namespace node from which this ProtoFile originated.
+   */
+  source: Namespace;
+}
+
+/**
+ * The built-in options that are defined by Protobuf.
+ */
+export interface WellKnownFileOptions {
+  java_package: string;
+  java_outer_classname: string;
+
+  optimize_for: "SPEED" | "CODE_SIZE" | "LITE_RUNTIME";
+
+  cc_enable_arenas: boolean;
+}
+
+/**
+ * A top-level declaration.
+ */
+export type ProtoTopLevelDeclaration =
+  | ProtoServiceDeclaration
+  | ProtoMessageDeclaration
+  | ProtoEnumDeclaration;
+
+/**
+ * A declaration. One of `service`, `message`, a field within a message, `one_of`, `enum`, or an `rpc` method.
+ */
+export type ProtoDeclaration =
+  | ProtoServiceDeclaration
+  | ProtoMessageDeclaration
+  | ProtoFieldDeclaration
+  | ProtoOneOfDeclaration
+  | ProtoEnumDeclaration
+  | ProtoMethodDeclaration;
+
+/**
+ * A Protobuf scalar type.
+ */
+export type ScalarName = ScalarIntegralName | "double" | "float" | "bytes" | "string";
+
+/**
+ * A Protobuf integral type.
+ */
+export type ScalarIntegralName = ScalarIntegerName | ScalarFixedName | "bool";
+
+/**
+ * A Protobuf variable-length integer type.
+ */
+export type ScalarIntegerName = `${"u" | "s" | ""}int${"32" | "64"}`;
+
+/**
+ * A Protobuf fixed-length integer type.
+ */
+export type ScalarFixedName = `${"s" | ""}fixed${"32" | "64"}`;
+
+// Symbols for type destructuring
+const $scalar = Symbol("$scalar");
```

Comment:
should those symbols be using `Symbol.For`?

Reply:
I think it's correct for these to be unique symbols rather than globally registered. I don't intend for anyone to ever construct or deconstruct these `ProtoType` data structures except through the functions in this module.

Reply:
it is mostly an issue if you have 2 version of the library loaded then they would have different symbols.

Reply:
And -- instances of the proto AST are not shared between instances of the emitter, so each copy of the lib would have its own copy of the AST data structure and its own copy of these internal symbols. If multiple versions of the library were loaded and the AST were shared, then it might be incompatible in other ways, so we can't assume that the proto AST can be shared anyways.